cli/sql: bump the go-libedit dependency to fix signal handling on darwin

knz opened the pull request:
The details of the story can be found here:

https://github.com/knz/go-libedit/commit/5ffa3044c44237e9fc907fd3fc17e4cf604a2f12
(failed attempt: https://github.com/knz/go-libedit/commit/3b0f69cb0f672aad6a49ee224c3a801ae576a6e9)

This patch also tweaks the handling of the prompts, since go-libedit
1.5 doesn't do Go callbacks any more.

Fixes #19132.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19429)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

```
  The second option seems "safer" (and also has the benefit
  of allowing the user to customize which key combo triggers
  the behavior) BUT doesn't work in practice: when libedit
  receives *another* signal, especially CONT (e.g. after
  a Ctrl+Z suspend), it will restore the terminal configuration
  to default and Ctrl+C will resume its standard behavior
  (send SIGINT). Thus Ctrl+C after a suspend will break things.
```

This sounds like a bug that should be fixed in libedit (or am I missing something?)

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19429#-:-Kx9E4ZLp103VFev4rJ3:b-qt9mhj)*
<!-- Sent from Reviewable.io -->

knz commented:
libedit does *not* handle ctrl+c as keyboard input; I was adding the code to do so in go-libedit by using the tcsetattr syscall. From libedit's perspective, 1) it has complete control over tcsetattr 2) the terminal should handle special signal-inducing combos.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19429#-:-Kx9FzdP0l73HVLEp4NN:b-vou3ej)*
<!-- Sent from Reviewable.io -->

knz commented:
Wow I'm super bummed with what TC just helped me realized: SIGINT may not be delivered on the same pthread as the go-libedit code, so using longjmp from the handler to the libedit code confuses the Go runtime (that can move the goroutine from one pthread to another behind its back). Of course since my small toy test was just using 1 goroutine all was well. 

Gosh, signals are hard. I'll try something else.

benesch commented:
Can you forward the signal via `pthread_kill` to the thread that called `sigsetjmp`?

knz commented:
> Can you forward the signal via pthread_kill to the thread that called sigsetjmp?

Interesting idea. I'll try next weekend. One more iteration! :)

knz commented:
mistake

knz commented:
I think I got it on second try. https://github.com/knz/go-libedit/commit/5ffa3044c44237e9fc907fd3fc17e4cf604a2f12 

yay

bdarnell commented:
LGTM

knz committed:
cli/sql: bump the go-libedit dependency to fix signal handling on darwin

The details of the story can be found here:

https://github.com/knz/go-libedit/commit/5ffa3044c44237e9fc907fd3fc17e4cf604a2f12

(previous attempt:
https://github.com/knz/go-libedit/commit/3b0f69cb0f672aad6a49ee224c3a801ae576a6e9)

This patch also tweaks the handling of the prompts, since go-libedit
1.5 doesn't do Go callbacks any more.

knz commented:
Thanks! :tada:
